### ISSUE TITLE
Torchscript execution failures (hangs, access violation, Fatal error. Internal CLR fatal error. (0x80131506) )

### ISSUE THREAD

**Issue by dje-dev** [dotnet/TorchSharp#1047]
Repeated execution of Torchscript modules leads to consistent failures of one of these varieties, apparently due to memory corruption (within about 30 seconds of runtime).

This has now been narrowed down to a fairly minimal repro. This hangs or gives CLR fatal error within about 20 seconds, repeatedly running a medium sized (26mm parameter) neural network and interleaving with exercising .NET garbage allocation.

This is running latest version of Torchsharp (0.100.3) on Windows 11 with CUDA 12.2. The TorchScript file is too large to easily upload (56MB) and possibly this would happen with any TorchScript module.

```
    public static void ReplicateFailure()
    {
      var module = torch.jit.load<Tensor, (Tensor, Tensor, Tensor, Tensor)>(@"c:\temp\fail.ts", DeviceType.CUDA, 0).to(ScalarType.Float16);

      for (int bs = 1; bs < 133; bs += 3)
      {
        Tensor input = torch.tensor(new float[bs * 64 * 135], ScalarType.Float16, new Device("cuda:0"), false)
                            .reshape(new long[] { bs, 64, 135 });
        for (int j = 0; j < 100; j++)
        {
          using (var dx = torch.NewDisposeScope())
          {
            module.call(input);
          }
          ExerciseGC(j);
        }
        Console.WriteLine(bs + " done loop this batch size ");
      }
    }

    public static void ExerciseGC(int index)
    {
      Console.WriteLine("in " + index);
      object[] objs = new object[20_000];
      for (int i = 0; i < 20_000; i += 999)
      {
        objs[i] = new byte[i * 10];
      }
      Console.WriteLine("out");
    }

```

**Comment by NiklasGustafsson** [comment]
Does it work better if you dispose the `input` tensor at the bottom of the loop? You're allocating a lot of GPU memory, as far as I can tell, and not disposing of it.

**Comment by dje-dev** [comment]
In the code, the single call to the Torchscript module is immediately enclosed with a NewDisposeScope(), I believe this will release the GPU memory every time thru the loop, correct?

**Comment by NiklasGustafsson** [comment]
> In the code, the single call to the Torchscript module is immediately enclosed with a NewDisposeScope(), I believe this will release the GPU memory every time thru the loop, correct?

It will release the memory produced by the module (i.e. its output) at the bottom of the inner loop, but not the `input` tensor that is allocated in the outer loop.

**Comment by dje-dev** [comment]
Yes, understood. I had missed the outer allocation.  Unfortunately the code you provide still hangs within 30 seconds. Is the next thing to try to replicate with a smaller Torchscript file that is accessible to both of us, and see if that can also replicate the problem? I have no problems at all with any of the non-Torchscript features of Torchsharp on my system.

**Comment by NiklasGustafsson** [comment]
Yes, it's hard to diagnose without more details. Unfortunately, I'm going to be out of the office until August 7th, but maybe we can continue this then?

**Comment by dje-dev** [comment]
Understood, thanks, I'll try to have data points to share upon your return.

**Comment by dje-dev** [comment]
Here's a more minimal repro of this problem that makes Torchscript not currently usable for me.

Repeated execution of unit test TestLoadJIT_3 will surface this problem, resulting in CLR errors, access violations and hangs.

For example, the following always fails within a minute or two of execution (on CPU, using explicit release of Tensor memory).

```
    int xx = 0;
    jit.ScriptModule<Tensor, Tensor> mx = torch.jit.load<Tensor, Tensor>("l1000_100_10.script.dat");
    while (true)
    {
      using var input = torch.ones(1000);
      using var output = mx.call(input);
      Console.WriteLine(xx++);
    }
```

Here's the tail output of one sample abend (after about 60 seconds):
```
242146
Fatal error. 242147
242148
Internal CLR error. (0x80131506)
242149
242150
```

**Comment by NiklasGustafsson** [comment]
And it works when you run it using PyTorch, presumably?

**Comment by NiklasGustafsson** [comment]
Interesting...

The problem is that when returning the result, the C++ code calls into managed code to allocate a 'TensorOrScalar' struct, which then has to be pinned in order to avoid being moved.

However, in IndexedPinnedArray.cs, line 44, the call to `AddrOfPinnedObject()` sometimes returns `IntPtr.Zero` for a valid, allocated handle to a pinned object. This will take some sleuthing to figure out.

**Comment by phineasng** [comment]
Hi all,

@NiklasGustafsson thanks for this great package. 

I was wondering if there is any update on this? Or if in the meantime somebody found out at least an hacky workaround.

**Comment by NiklasGustafsson** [comment]
I come back to this one every now and then when I'm bored... :-) Still not any closer to figuring out what's going on.

**Comment by dje-dev** [comment]
Thank you for the update! I was also going to ask, this is definitely an impediment for my work too.

I'm sorry it's proving so much a mystery. Here's one radical thought - maybe you could try upgrading to more recent Pytorch (2.1) and CUDA (12.x) versions. This could have two benefits, it might fix the problem if it was an old bug or at least behave differently and give more clues. Additionally, a package upgrade would be very helpful (especially since the current one does not support modern hardware such as H100,  reporting "no kernel found for device" when Torchsharp is run.

**Comment by NiklasGustafsson** [comment]
Upgrading to 2.1 isn't that radical, per se, but the move to CUDA 11.8 or 12 that it requires will be a lot of work (we're on 11.7 right now). It's something that needs to be done, but it will take some preparation and a few solid days of dedicated time.

It also requires help from someone who has the authority to take ownership on behalf of the .NET Foundation, because every CUDA version change requires a package name change, and that requires ownership transfer and some coordination.

**Comment by dje-dev** [comment]
Understood, good luck. CUDA 11.7 is only 14 months old but already feels ancient! For example, Pytorch 2.1 installation page seems to indicate that the minimum requirement is CUDA 11.8.

**Comment by NiklasGustafsson** [comment]
I got started on the move, but I'm running into the 'couldn't load nvfuser_codegen' warning on Windows again, which I thought had been addressed, so that will take some figuring out. Also, it'll be important to be on the latest and greatest NVIDIA drivers.

**Comment by dje-dev** [comment]
Great news. It's true we get this nvfuser error too, but it seems benign as far as I can tell. And looking online, it seems likely that this nvfuser is or has been deprecated in Pytorch.

**Comment by NiklasGustafsson** [comment]
It's still there in 2.1.0, and it's not benign -- it seems to disable the CUDA backend when it fails to load. Either that, or there's some other issue that disables it. I hope it's the same problem -- I'd rather solve one than two.

**Comment by dje-dev** [comment]
I can definitively report that in my environment (Windows) I get this warning/error but CUDA definitely still works, so you perhaps have a different cause.

[W C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\jit\codegen\cuda\interface.cpp:47] Warning: Loading nvfuser library failed with: error in LoadLibrary for nvfuser_codegen.dll. WinError 126: The specified module could not be found.
 (function LoadingNvfuserLibrary)

 Pytorch version  : 2.0.1.1
  Training device : cuda:0

**Comment by NiklasGustafsson** [comment]
Installing the CUDA 12.1 Toolkit seems to take care of both problems.

**Comment by NiklasGustafsson** [comment]
Status update on 2.1.0 work: I have it working on Windows, with CUDA, but I'm having a challenging time getting the Linux and MacOS builds to build. libtorch has moved to C++ 17, and enabling it creates problems because CLANG (unlike MSVC) doesn't simply ignore the command line option for the one file that is C rather than C++, so it rejects it.

The problem is in the src/Native/CMakeList.txt file, where I don't know how to add the compiler option for just C++, excluding C.

**Comment by dje-dev** [comment]
Good news! 

Regarding your challenge, GPT4 produced a plausible solution, but I'm not qualified to judge if it's just a hallucination..

.................
Building cross-platform C++ projects can indeed be challenging due to differences in compiler behavior and available features. To address your issue with setting compiler options for C++ files only, you can use the $<COMPILE_LANGUAGE:CXX> generator expression in CMake. This expression evaluates to true for C++ sources.

You can set the C++ standard for your target using the target_compile_features function, and for setting specific compiler options, you can use target_compile_options. Here is an example of how you might adjust your CMakeLists.txt:

cmake
```
if (WIN32)
    [...]
else()
    add_definitions(-Werror) # treat warnings as errors
    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
    add_compile_options(-Wno-unused-local-typedef)
    add_compile_options(-fPIC)
    add_compile_options(-fvisibility=hidden)

    # Set C++ standard
    set_target_properties(your_target_name PROPERTIES
        CXX_STANDARD 17
        CXX_STANDARD_REQUIRED ON
    )

    # Add C++ specific compile options
    target_compile_options(your_target_name PRIVATE
        $<$<COMPILE_LANGUAGE:CXX>:-std=c++17>
    )
endif()
```
In this example:

Replace your_target_name with the actual name of your target.
CXX_STANDARD is used to set the required C++ standard.
CXX_STANDARD_REQUIRED ensures that CMake will fail configuration if the required standard is not supported.
target_compile_options is used to add compile options conditionally for C++ files.
These commands ensure that the C++ standard and compiler options are applied only to C++ files, not to C files. Adjust the options as needed for your project’s requirements.

**Comment by NiklasGustafsson** [comment]
So, v0.101.0, with libtorch 2.1.0 support is out. I have not checked whether it addresses the problem you reported in this bug.

**Comment by NiklasGustafsson** [comment]
I can confirm that this problem persists. I will continue to look into it. I understand that it's blocking.

**Comment by NiklasGustafsson** [comment]
It makes me wonder if the method we're using to pass data between TorchScript and .NET is too clever -- calling back to .NET to allocate arrays of TensorOrScalar, instead of allocating them in native code and then freeing up.

**Comment by NiklasGustafsson** [comment]
Okay, so by allocating and freeing memory in native code, instead of pinning managed objects, it both faster and can run your repro case (and a few others) 1,000,000 times. 

It takes a very long time (15 minutes) to do so, so I'm going to have to use a lower number of iterations for the checked-in unit test.

**Comment by dje-dev** [comment]
Really great news on both the 2.1.0 upgrade and this fix, many thanks.

**Comment by NiklasGustafsson** [comment]
Since this is a really significant bug to have fixed, I will make another release just for this.

**Comment by NiklasGustafsson** [comment]
Addressed in v0.101.1, just release to NuGet Gallery.
